## Importing the Beautiful soup

In [1]:
from bs4 import BeautifulSoup
import numpy as np
from urllib.request import urlopen
import requests
import pandas as pd

# Building Source code

In [2]:
source = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page = urlopen(source).read().decode('utf-8')
soup = BeautifulSoup(page,'html.parser')
wiki_table = soup.body.table.tbody

# Extracting Data

In [3]:
def get_cell(element):
    cells = element.find_all('td')
    row = []
    
    for cell in cells:
        if cell.a:            
            if (cell.a.text):
                row.append(cell.a.text)
                continue
        row.append(cell.string.strip())
        
    return row

print('get_cell Defined')

get_cell Defined


In [4]:
def get_row():    
    data = []  
    
    for tr in wiki_table.find_all('tr'):
        row = get_cell(tr)
        if len(row) != 3:
            continue
        data.append(row)        
    
    return data

# Getting the Data

In [5]:
data = get_row()
columns = ['Postcode', 'Borough', 'Neighbourhood']
df = pd.DataFrame(data, columns=columns)
df.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [6]:
df.shape

(288, 3)

# Clinning the Data

In [7]:
df1 = df[df.Borough != 'Not assigned']
df1 = df1.sort_values(by=['Postcode','Borough'])

df1.reset_index(inplace=True)
df1.drop('index',axis=1,inplace=True)

df1.head()

Postcode      Borough   Neighbourhood
0      M1B  Scarborough           Rouge
1      M1B  Scarborough         Malvern
2      M1C  Scarborough  Highland Creek
3      M1C  Scarborough      Rouge Hill
4      M1C  Scarborough      Port Union

In [8]:
df_postcodes = df1['Postcode']
df_postcodes.drop_duplicates(inplace=True)
df2 = pd.DataFrame(df_postcodes)
df2['Borough'] = '';
df2['Neighbourhood'] = '';


df2.reset_index(inplace=True)
df2.drop('index', axis=1, inplace=True)
df1.reset_index(inplace=True)
df1.drop('index', axis=1, inplace=True)

for i in df2.index:
    for j in df1.index:
        if df2.iloc[i, 0] == df1.iloc[j, 0]:
            df2.iloc[i, 1] = df1.iloc[j, 1]
            df2.iloc[i, 2] = df2.iloc[i, 2] + ',' + df1.iloc[j, 2]
            
for i in df2.index:
    s = df2.iloc[i, 2]
    if s[0] == ',':
        s =s [1:]
    df2.iloc[i,2 ] = s

In [9]:
df2.shape

(103, 3)

## Here our data contains 103 rows

# Week 03 No 2 Adding geographical Data

## Importing required library

In [10]:
import json
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
#!conda install -c conda-forge folium=0.5.0 --yes
import folium

In [11]:
df2['Latitude'] = '0';
df2['Longitude'] = '0';

df_geo = pd.read_csv('https://cocl.us/Geospatial_data')
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [12]:
for i in df2.index:
    for j in df_geo.index:
        if df2.iloc[i, 0] == df_geo.iloc[j, 0]:
            df2.iloc[i, 3] = df_geo.iloc[j, 1]
            df2.iloc[i, 4] = df_geo.iloc[j, 2]
            
df2.head()

Postcode      Borough                         Neighbourhood Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.8067   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.7845   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.7636   
3      M1G  Scarborough                                Woburn   43.771   
4      M1H  Scarborough                             Cedarbrae  43.7731   

  Longitude  
0  -79.1944  
1  -79.1605  
2  -79.1887  
3  -79.2169  
4  -79.2395